src/walker.rs

```diff
@@ -0,0 +1,246 @@
+#![allow(missing_docs)]
+
+use std::collections::HashSet;
+
+use derive_more::{Deref, DerefMut};
+use itertools::Itertools;
+
+use crate::{ops::OpType, HugrView, Node};
+
+#[derive(Eq, PartialEq, Ord, PartialOrd, Copy, Clone, Debug)]
+pub enum WalkResult {
+    Advance,
+    Interrupt,
+}
+
+#[derive(Eq, PartialEq, Ord, PartialOrd, Copy, Clone, Debug)]
+pub enum WalkOrder {
+    Preorder,
+    Postorder,
+}
+
+#[derive(Deref, DerefMut)]
+struct WalkerCallback<'a, T, E>(Box<dyn 'a + FnMut(Node, OpType, T) -> Result<T, E>>);
+
+impl<'a, T, E, F: 'a + FnMut(Node, OpType, T) -> Result<T, E>> From<F>
+    for WalkerCallback<'a, T, E>
+{
+    fn from(f: F) -> Self {
+        Self(Box::new(f))
+    }
+}
+
+pub struct Walker<'a, H: HugrView, T, E> {
+    pre_callbacks: Vec<WalkerCallback<'a, T, E>>,
+    post_callbacks: Vec<WalkerCallback<'a, T, E>>,
+    hugr: &'a H,
+}
+
+enum WorkItem {
+    Visit(Node),
+    Callback(WalkOrder, Node),
+}
+
+impl<'a, H: HugrView, T, E> Walker<'a, H, T, E> {
+    pub fn new(hugr: &'a H) -> Self {
+        Self {
+            pre_callbacks: Vec::new(),
+            post_callbacks: Vec::new(),
+            hugr,
+        }
+    }
+
+    pub fn previsit<O, F: 'a + FnMut(Node, O, T) -> Result<T, E>>(&mut self, f: F) -> &mut Self
+    where
+        OpType: TryInto<O>,
+    {
+        self.visit(WalkOrder::Preorder, f)
+    }
+
+    pub fn postvisit<O, F: 'a + FnMut(Node, O, T) -> Result<T, E>>(&mut self, f: F) -> &mut Self
+    where
+        OpType: TryInto<O>,
+    {
+        self.visit(WalkOrder::Postorder, f)
+    }
+
+    fn mut_callbacks(&mut self, order: WalkOrder) -> &mut Vec<WalkerCallback<'a, T, E>> {
+        match order {
+            WalkOrder::Preorder => &mut self.pre_callbacks,
+            WalkOrder::Postorder => &mut self.post_callbacks,
+        }
+    }
+
+    pub fn visit<O, F: 'a + FnMut(Node, O, T) -> Result<T, E>>(
+        &mut self,
+        walk_order: WalkOrder,
+        mut f: F,
+    ) -> &mut Self
+    where
+        OpType: TryInto<O>,
+    {
+        let cb = move |n, o: OpType, t| match o.try_into() {
+            Ok(x) => f(n, x, t),
+            _ => Ok(t),
+        };
+        self.mut_callbacks(walk_order).push(cb.into());
+        self
+    }
+
+    pub fn walk(&mut self, mut t: T) -> Result<T, E> {
```

Comment:
Because the callbacks need access to the hugr. In the current design the hugr is fixed for a given Walker, so the callbacks can capture the hugr. Passing the hugr here would mean that the Walker could be invoked on many hugrs, so the callbacks couldn't capture it. The hugr would need to be threaded through the callbacks. This can be done but I think it's less ergonomic.